### PR TITLE
i2c: Promote use of Deref trait

### PR DESCRIPTION
**Pull request by keisrk**
Modeled after stm32f4xx-hal's I2C implementation. The pattern is
widely seen in other HAL families as well.

### CODE REVIEW

**Comment by keisrk** [comment]
This PR might help the other issue [Missing I2C3 Peripheral on STM32L476 #111](https://github.com/stm32-rs/stm32l4xx-hal/issues/111).

File: src/i2c.rs [hunk 1]
```diff
@@ -74,253 +74,275 @@ macro_rules! busy_wait {
...
+        clocks: Clocks,
+    ) -> Self where
+        F: Into<Hertz>,
+        SCL: SclPin<I2C>,
+        SDA: SdaPin<I2C>,
+    {
+        let freq = freq.into().0;
+
+        assert!(freq <= 1_000_000);
+
```
Comment (MathiasKoch) [inline src/i2c.rs:131]:
```suggestion
        // Make sure the I2C unit is disabled so we can configure it
        i2c.cr1.modify(|_, w| w.pe().clear_bit());
```

**Review by MathiasKoch — APPROVED** [review]
I like it very much! 
Much better to read, understand and lint than the excessive marco_rules

Added one small suggestion.

File: src/i2c.rs (hunk 1 above)
Comment (keisrk) [reply src/i2c.rs:131]: Sure I added the code.

**Comment by keisrk** [comment]
Thanks @MathiasKoch I reflected your comment and merged it into our dev branch.

**Pull request merged**